contracts/VolOracle.sol

```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity >=0.8.4;
+
+import { Observation } from "@uniswap/v3-core/contracts/libraries/Oracle.sol";
+import { IUniswapV3Pool } from "@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol";
+
+contract VolOracle {
+    // constructor
+    // initializes a uniswap
+
+    // observation size
+    // stores 30 days of data (with some buffer)
+    uint256 public constant OBSERVATION_SIZE = 345600;
+    uint256 public constant UNIV3_MAX_CARDINALITY = 65535;
+    // @dev Stores Observation arrays for each pool
+    mapping(address => Observation[UNIV3_MAX_CARDINALITY]) public observations;
+    // @dev Stores lastBlockTimestamp when the observation was initialized for the pool
+    mapping(address => uint256) public lastBlockTimestamps;
+    mapping(address => uint256) public lastObservationIndex;
+
+    function initPool(address _pool) external {
+        require(observations[_pool].length == 0, "Pool already initialized");
+        // only initialize pools which have max cardinality
+        require(
+            IUniswapV3Pool(_pool).slot0.observationCardinality == UNIV3_MAX_CARDINALITY,
+            "Pool not at max cardinality"
+        );
+        // TODO: if pool is not at max cardinality then grow it
+        // initializing the pool to max size
+        observations[_pool] = new Observation[](OBSERVATION_SIZE);
```

Comment:
I tried locally, but seems this kind of initialization only works for raw primitive array like uint256, doing this on struct will get: UnimplementedFeatureError: Copying of type struct memory[] memory to storage not yet supported.

Reply:
we can try something like this: https://github.com/Uniswap/v3-core/blob/main/contracts/UniswapV3Pool.sol#L99